### ISSUE TITLE
Enter-PSSession Error processing data from the background process

### ISSUE THREAD

**Issue by veleek** [PowerShell/PowerShell#12852]
SYMPTOMS
After successfully configuring OpenSSH as a server on Windows, confirming that access with ssh.exe or putty.exe works as expected, when connecting with pwsh's Enter-PSSession or Invoke-Command, it fails with an error message: "There is an error processing data from the background process. Error reported: ..." When you delete, rename or move your PowerShell profile script(s) at that OpenSSH server, the error message disappears and pwsh remoting works as expected.

CAUSE
You have 1) a profile script at the OpenSSH server, 2) that profile script prints or displays output when run, and 3) you have changed the default OpenSSH server's command shell from CMD to either Windows PowerShell or PowerShell Core. The output of the profile script is disrupting the launch of pwsh as an OpenSSH "subsystem" despite the fact that the sshd_config file is correctly configured with the "..\pwsh.exe -sshs -nologo -noprofile" line.

REPRODUCE
Install PowerShell Core 7 preview 5 on a Windows Server 2019 system.

Delete, rename or move all PowerShell profile scripts on that server.

Install the OpenSSH 8.0.beta sshd.exe service from GitHub on that Windows Server into $env:ProgramFiles\OpenSSH as recommended, and configure the "Subsystem" line in it's sshd_config file (https://docs.microsoft.com/en-us/powershell/scripting/learn/remoting/ssh-remoting-in-powershell-core?view=powershell-6).

Confirm that the OpenSSH server is working with both ssh.exe and pwsh's "Enter-PSSession -HostName server."

Change the OpenSSH default command shell to either powershell.exe or pwsh.exe (https://github.com/PowerShell/Win32-OpenSSH/wiki/DefaultShell).

Confirm with ssh.exe and "Enter-PSSession -HostName server" that OpenSSH is still accessible and that ssh.exe gets the new default shell as expected (powershell.exe or pwsh.exe, not CMD).

Add a profile script for powershell.exe or pwsh.exe, whichever one you want, and add a command to that script which produces output, e.g., a line like "XXXXXXX".

Confirm with ssh.exe that OpenSSH is still accessible and that the profile script runs as expected, e.g., it prints "XXXXXXX."

Attempt to connect with pwsh's "Enter-PSSession -HostName server" as before. Note that it now fails with an error message: "There is an error processing data from the background process. Error reported: XXXXXXX."

Comment out the line from the profile script which produces the output. Add other commands to the profile script which do not produce output, e.g., create a function. Attempt to connect again with pwsh's "Enter-PSSession -HostName server" and note that it succeeds.

Uncomment the line from the profile script which produces the output and causes pwsh's Enter-PSSession to fail.

Change the default OpenSSH command shell back to CMD again.

Attempt to connect again with pwsh's "Enter-PSSession -HostName server" and note that it succeeds.

NOTES
Set the default OpenSSH command shell back to CMD. Use Process Explorer or a similar tool to watch the sshd.exe service as a pwsh client connects with "Enter-PSSession -HostName server". The parent sshd.exe spawns a child sshd.exe, which runs "cmd.exe /c" to launch "pwsh.exe -sshs -nologo -noprofile". It works.

But when 1) the default OpenSSH shell is changed to either powershell.exe or pwsh.exe and 2) there is a profile script that produces output, then the parent sshd.exe runs either "powershell.exe -c" or "pwsh.exe -c" (whichever you chose) in order to launch "pwsh.exe -sshs -nologo -noprofile". The profile script of powershell.exe/pwsh.exe is disrupting the launch of the last pwsh.exe in the chain of process launches.

The "-c" argument to powershell.exe/pwsh.exe seems to be hard-coded into sshd.exe. Following the above instructions for changing the default OpenSSH command shell, if you configure the DefaultShellCommandOption registry value, this value appears to be ignored by sshd.exe when launching pwsh as a subsystem. (See https://github.com/PowerShell/Win32-OpenSSH/wiki/DefaultShell.) Setting the DefaultShellCommandOption value to something like "-noprofile -command" does not work.

This issue has not been tested on a Linux SSH server.

This issue has not been tested on any earlier versions of pwsh or OpenSSH for Windows.

EXPECTED BEHAVIOR
When the default OpenSSH command shell is changed from CMD to either powershell.exe or pwsh.exe on a Windows server, and the SSH user has a profile script on that server, then that profile script should not be run when the user connects with "Enter-PSSession -HostName server" using pwsh.

MITIGATION
Either 1) change the default OpenSSH command shell back to CMD on the target server, or 2) edit your profile script on the server so that it produces no output when it runs. The profile script can create functions, change the Prompt function, change colors, etc., just so long as it produces no output which confuses sshd.exe.

NOTE: This is a re-creation of https://github.com/PowerShell/Win32-OpenSSH/issues/1498 created at the direction of @bagajjal and originally created by @JasonFossen.

**Comment by yobyot** [comment]
Sheesh...entering a simple `$PSVersionTable` in `$Profile` to make sure one knows which version of PowerShell is being invoked breaks `New-PsSession`. And, as usual, _it's not doc'd anywhere_. Without @veleek 's detailed problem statement, I'd be flinging insults (again) at ssh on Windows.

**Comment by veleek** [comment]
> I used Windows11 to connect to debain 11 and had this problem
> 
> ![image](https://user-images.githubusercontent.com/37170330/211558925-0ccf1ce8-5fdb-4c6e-b716-0d20e70b13e4.png)
> 
> ![image](https://user-images.githubusercontent.com/37170330/211559079-e7aed6b1-7653-431c-91fc-82484a00ca8d.png)

Yes, the text after `Error Reported` is the output from the powershell profile (or some other source when powershell starts) that is incorrectly being interpreted as an error.  If you can figure out where it's coming from and comment it out or disable it, then you should be fine.

**Comment by veleek** [comment]
> How do I find the specific error message?Is there a reference document?

It's not an actual error.  That's what this GitHub issue is talking about.  Your PowerShell setup is outputting the text "PowerShell 7.3.1" during regular operation and it's (incorrectly) being interpreted as an error causing the remote session to fail.  

Because it's not an error and is specific to your setup there is not an easy way for us to tell you how to fix the problem.  You need to look through the config and profiles for both machines and figure out what is writing that specific text and change or disable it.

**Comment by veleek** [comment]
I cannot help you with this problem.  You need to look at the powershell profile or SSH configuration on the Linux environment you are connecting to and determine what is outputting the text.

**Comment by MFunction96** [comment]
Still Powershell 7.3.1, the latest stable version. 
![image](https://user-images.githubusercontent.com/19974616/214035294-bb134e38-2acd-4d42-947e-6fafb56185dc.png)
I had met the same issue when I tried to connect Linux via SSH. The `Powershell 7.3.1`, which output is ahead all prompts, makes New-PSSession treat the output as some kind of error when starting connection. But it is normal output when starting powershell.
![image](https://user-images.githubusercontent.com/19974616/214035735-3a7bc520-d393-4fac-9892-308bcbcc43d5.png)

**Comment by veleek** [comment]
That version number is being output by PowerShell as part of the "header logo".  If you run `pwsh` using the `--NoLogo` (or `-nol`) parameter it won't output the version.

```
# Goes directly to new prompt no version
~> pwsh --NoProfile --NoLogo
PS C:\Users\veleek> exit

# Outputs version number prior to prompt
~ >pwsh --NoProfile
PowerShell 7.3.1
PS C:\Users\veleek>
```

I'm not aware of any way to pass additional arguments to the remote powershell session, so it looks like this just needs to be fixed by the powershell team.

**Comment by MFunction96** [comment]
> 

Thanks for your reply. I tried it and worked. It seems there is a conflict between the actual and the [document](https://learn.microsoft.com/en-us/powershell/scripting/learn/remoting/ssh-remoting-in-powershell-core?view=powershell-7.3#install-the-ssh-service-on-a-windows-computer). Whatever, the problem was solved.

![image](https://user-images.githubusercontent.com/19974616/214181914-573d226f-247f-4c8a-ba1e-449fea3d2ab5.png)

**Comment by veleek** [comment]
@MFunction96 - Dope!  I'm glad there's an option to pass `-nologo` that works!  I [submitted a fix to remove that note](https://github.com/MicrosoftDocs/PowerShell-Docs/pull/9745) and pinged the original author.  We'll see if it's intended or if there's some other configuration we're missing that's causing problems.

**Comment by sdwheeler** [comment]
@veleek There was a miscommunication from the PowerShell team. The fix for this is in 7.4-preview.1, not 7.3.

**Comment by veleek** [comment]
@coolqingcheng, follow the [instructions here](https://learn.microsoft.com/en-us/powershell/scripting/learn/remoting/ssh-remoting-in-powershell-core?view=powershell-7.3#install-the-ssh-service-on-a-windows-computer) but ignore the note about not needing `-nologo`.
